### ISSUE TITLE
Referencing Julia object from generated code

### ISSUE THREAD

**Issue by yuyichao** [JuliaLang/julia#11695]
- TL;DR;
  
  How should we handle direct reference of julia object in generated code that is not visible to GC otherwise.
- SegFault example
  
  This example is specifically constructed (i.e. it is written with crashing julia in mind) to demonstrate the more general problem in an explicit way. Although I am more interested in a special case, which is also more implicit (See below).
  
  ``` julia
  function f()
      m = Module(:m)
      finalizer(m, (m)->ccall(:jl_, Void, (Any,), "Finalizing m"))
      a = eval(m, :(a(::Int) = m))
      b = cfunction(a, Any, (Int,))
      @eval function g()
          ccall($b, Any, (Int,), 1)
      end
  end
  
  f()
  
  gc()
  println(g())
  gc()
  println(g())
  ```
- The problem (To the best of my knowledge)
  
  The general problem is that there's no reference from the generated code to julia objects and we sometimes embed pointers to julia objects directly in the code. There are (at least) two cases where this can happen,
  1. (The case above) When the AST has objects embedded in them (e.g. because of metaprogramming). These values seems to be referenced by the AST but can be out of reach if the function itself is collected. (Actually, I couldn't figure out how are they referenced especially when the AST is compressed. Would be nice to have a GC visualizer =) )
  2. (What I'm more interested in) We embed the function object into the AST during codegen if we can statically evaluate them at compile time, especially for `jlcall` signature. This is currently not an issue because non of the generated code is actually using them. However, for a few of my PR's and more fine grain method caching in general (e.g. @carnaval 's JSoC idea, which is the reason I bring it up now), it might be necessary to access these values.

For the proof of concept implementation of some codegen optimization I'm interested in, I might just keep a global array but I would like to know if there's better solution to this.

P.S. IMHO, the segfault in the example above should either be fixed or documented.

**Comment by yuyichao** [comment]
I'm asking also because I think this is necessary for static compilation and there might already be a solution for it? @vtjnash ?

**Comment by yuyichao** [comment]
Oh. And the output of the example above. It should be pretty obvious why it could segfault but anyway, here's the proof that it does segfault in case someone don't believe it ......

```julia
"Finalizing m"
m

signal (11): 段错误
jl_object_id at /usr/bin/../lib/julia/libjulia.so (unknown line)
unknown function (ip: -156319479)
unknown function (ip: -156310121)
unknown function (ip: -156265908)
jl_apply_generic at /usr/bin/../lib/julia/libjulia.so (unknown line)
unknown function (ip: -155908621)
unknown function (ip: -155911992)
unknown function (ip: -155831178)
unknown function (ip: -155828124)
jl_load at /usr/bin/../lib/julia/libjulia.so (unknown line)
unknown function (ip: -201927872)
jl_apply_generic at /usr/bin/../lib/julia/libjulia.so (unknown line)
include_from_node1 at ./loading.jl:133
jl_apply_generic at /usr/bin/../lib/julia/libjulia.so (unknown line)
unknown function (ip: -200036443)
unknown function (ip: -200041752)
unknown function (ip: -200038999)
jl_apply_generic at /usr/bin/../lib/julia/libjulia.so (unknown line)
unknown function (ip: 4201207)
unknown function (ip: 4200111)
__libc_start_main at /usr/bin/../lib/libc.so.6 (unknown line)
unknown function (ip: 4200201)
unknown function (ip: 0)
```

**Comment by JeffBezanson** [comment]
Normally roots in code are kept in the `roots` array of a `LambdaStaticData`. The problem here is that the only reference to the function is via `cfunction`, which returns a bare pointer. This is just like calling `pointer` on an array, which can lead to segfaults for the same reason.

**Comment by yuyichao** [comment]
Yes, for the example above it is the problem.
I would argue that one small difference is that for `pointer` on an array, we free the pointer but for the function case, we never free the compiled function.

**Comment by gbaraldi** [comment]
I tried reproducing this with the following code
```julia
function f()
    m = Module(:m)
    finalizer(m) do m
        ccall(:jl_, Cvoid, (Any,), "Finalizing m")
    end
    a = Base.eval(m, :(a(::Int) = m))
    @show a
    b = @cfunction($a, Any, (Int,))
    @show b
    @eval function g()
        ccall($b.ptr, Any, (Int,), 1)
    end
end

f()

GC.gc()
println(g())
GC.gc()
println(g())
```
And got no issues, it seems creating the `@cfunction` closure roots stuff

**Comment by vtjnash** [comment]
Accessing fields of cfunction as you did is solidly UB, so we can probably close this

**Comment by vtjnash** [comment]
Very few of the objects in the original example still exist in their original form, because of this issue and others related to it
